cp1616/include/cp1616/cp1616_io_controller.h

```diff
@@ -0,0 +1,278 @@
+   *
+   * \param mode - current mode
+   */
+          void setCpCurrentModeFlag(PNIO_MODE_TYPE mode);
+
+  /**
+   * \brief Returns IO Controller cp_current_mode_ flag
+   *
+   * \return cp_current_mode_
+   */
+          PNIO_MODE_TYPE getCpCurrentModeFlag();
+
+  /**
+   * \brief Sets sem_mode_change_ flag (used in callbackForModeChangeIndication)
+   *
+   * \param value
+   */
+          void setSemModChange(int mod_change);
+
+  /** Debugging functions  */
+  void printOutputData(unsigned int module);
+  void printInputData(unsigned int module);
+
+private:
+
+  /**
+   * \brief Constructor
+   */
+         Cp1616IOController();
+
+  /**
+   * \brief Static instance pointer
+   */
+         static Cp1616IOController *controller_instance_;
+
+  /**
+   * \brief CP ID according to TIA portal setup (default = 1)
+   */
+          PNIO_UINT32 cp_id_;
+
+  /**
+   * brief CP Handle obtained by PNIO_controller_open function
+   */
+          PNIO_UINT32 cp_handle_;
+
+  /**
+   * \brief CP ready communication flag
+   */
+          int cp_ready_;
+
+  /**
+   * \brief CP current mode (see PGH_IO-Base_76.pdf )
+   */
+          /*volatile*/ PNIO_MODE_TYPE cp_current_mode_;
+
+   /**
+   * \brief CP local state obtained by PNIO_data_write/PNIO_data_read functions
+   */
+          /*volatile*/ PNIO_IOXS cp_local_state_;
+
+   /**
+    * \brief flag used by CallbackForModeChangeIndication
+    */
+          int sem_mod_change_;
+
+  /**
+   * \brief counter of active input modules
+   */
+          PNIO_UINT32 device_input_count_;
+
+  /**
+   * \brief array of input module data lenghts
+   */
+          PNIO_UINT32 *device_input_length_;
+
+  /**
+   * \brief array of device input states
+   */
+          PNIO_IOXS /*volatile* volatile*/ *device_input_state_;
+
+  /**
+   * \brief array of device input addresses
+   */
+          PNIO_ADDR *device_input_address_;
+
+
+  /**
+   * \brief counter of active output modules
+   */
+          PNIO_UINT32 device_output_count_;
+
+  /**
+   * \brief array of output module data lenghts
+   */
+          PNIO_UINT32 *device_output_length_;
+
+  /**
+   * \brief array of device output states
+   */
+          PNIO_IOXS /*volatile* volatile*/ *device_output_state_;
+
+  /**
+   * \brief array of device output addresses
+   */
+          PNIO_ADDR *device_output_address_;
+
+  /**
+   * \brief array of input module data
+   */
+          PNIO_UINT8 *in_module_data_;
+
+  /**
+   * \brief array of arrays of input module data
+   */
+          PNIO_UINT8 **in_data_;
+
+  /**
+   * \brief total size of input data
+   */
+          unsigned int total_input_size_;
+
+  /**
+   * \brief array of output module data
+   */
+          PNIO_UINT8 *out_module_data_;
+
+  /**
+   * \brief array of arrays of output module data
+   */
+          PNIO_UINT8 **out_data_;
+
+  /**
+   * \brief total size of output data
+   */
+          unsigned int total_output_size_;
+
```

Comment:
There are too many pointers in this class.  This is normally a recipe for memory leaks.  I'll see what I find when I review the cpp?